lib/mastermind.rb

```diff
@@ -1,40 +1,122 @@
 class Mastermind
-  attr_reader :red_count, :white_count
+  attr_reader :restart
 
-  def print_instructions
+  def initialize(code, input)
+    @game_code = code
+    @game_input = input
+    @remaining_guesses = 10
+    @past_guesses = []
+    @restart = false
+  end
+
+  def print_instructions(long_colors)
     puts "\tWelcome to Mastermind! The object of the game is to guess a 4-color code
-        within 10 guesses. The available colors are red, orange, yellow, green,
-        blue, and purple. Colors may be used in the code more than once or not at all.
+        (order matters) within 10 guesses. The available colors are #{long_colors}.
+        Colors may be used in the code more than once or not at all.
 
         To make a guess: type the first letter of each color you're guessing in caps.
-        For example, if you want to guess 'red, orange, yellow, red,' type 'ROYR'.
+        For example, if you want to guess red, orange, yellow, red, enter 'ROYR'.
 
         After each guess, you will receive feedback in the form of colored pins,
         letting you know:
           \t- how many colors are correct but in the wrong position (white pins)
           \t- how many colors are correct and in the right position (red pins)
 
-        To quit the game, type '/q'. To start a new game, type '/r'."
+        To quit the game, type 'quit'. To start a new game, type 'restart'."
+  end
+
+  def loop_guesses(short_colors)
+    feedback = ""
+    while @remaining_guesses > 0
+      puts "You have #{@remaining_guesses} guesses remaining."
+      guess = @game_input.guess(short_colors)
+      quit if guess == "QUIT"
+      if guess == "RESTART"
```

Comment:
Looks like we're duplicating some of the logic inside `InputGuess` class. Maybe we can create a `Guess` class that can take the user input, validate it, and respond to `restart?`, `quit?`, and any other stuff you'd need to find out about the user's guess

Reply:
Agreed. Would it work to move quit and restart functionality to `InputGuess` rather than create a new class? It's technically not a guess being input but I could generalize the `InputGuess` class to be `UserInput`.

Reply:
You're definitely on the right track here, but I'd like to see it in a simple class on it's own. In object-oriented programming, we create objects with either data or behavior and we pass these objects around to other objects. Since each object has it's own set of behavior, we can encapsulate logic that pertains to a certain object inside a method on that object.

So a class like `InputGuess` can be a little off-putting because it represents a verb, not a noun. A more object oriented approach might be something like the following:
```rb
receiver = InputReceiver.new
text = receiver.get_input! # .chomp and other immediate formatting can be encapsulated in this guy
guess = Guess.new(text) # now we can ask the guess whatever we want to know about it
guess.quit? # false
guess.restart? # false
guess.to_s # "abcd"
guess.to_a # ["a", "b", "c", "d"]
result = GuessResult.new(guess, some_secret_code)
result.correct? # false
result.incorrect? # true
```
An ideal OO application would have all of the business/game logic encapsulated inside different, yet straightforward objects which get passed around to other objects. Notice in the example how `GuessResult` takes a `Guess` object. The `GuessResult` can use the convenience methods on the `Guess` object to determine whether it is correct or not! 


We can work on this more tomorrow if this didn't make sense!